okhttp/src/main/java/com/squareup/okhttp/ConnectionSpec.java

```diff
@@ -143,35 +167,65 @@ void apply(SSLSocket sslSocket, Route route) {
+    return new Builder(this)
+        .cipherSuites(cipherSuitesToEnable)
+        .tlsVersions(protocolsToEnable)
+        .build();
   }
 
   /**
-   * Returns a copy of this that omits cipher suites and TLS versions not
-   * enabled by {@code sslSocket}.
+   * Returns {@code true} if the socket, as currently configured, supports this ConnectionSpec.
+   * In order for a socket to be compatible the enabled cipher suites and protocols must match.
+   *
+   * <p>For cipher suites, at least one of the {@link #cipherSuites() required cipher suites} must
+   * match the socket's enabled cipher suites. If there are no required cipher suites the socket
+   * must have at least one cipher suite enable.
+   *
+   * <p>For protocols, the first {@link #tlsVersions() TLS version} is considered a minimum
+   * requirement and must be be present in the socket's enabled protocols. For example, a
+   * connection spec containing [TLSv1.2, TLSv1.1, etc.] is only compatible if the socket has
+   * TLSv1.2 enabled.
    */
-  private ConnectionSpec supportedSpec(SSLSocket sslSocket) {
-    String[] cipherSuitesToEnable = null;
-    if (cipherSuites != null) {
-      String[] cipherSuitesToSelectFrom = sslSocket.getEnabledCipherSuites();
-      cipherSuitesToEnable =
-          Util.intersect(String.class, cipherSuites, cipherSuitesToSelectFrom);
+  public boolean isCompatible(SSLSocket socket) {
+    if (!tls) {
+      return false;
     }
 
-    String[] protocolsToSelectFrom = sslSocket.getEnabledProtocols();
-    String[] tlsVersionsToEnable = Util.intersect(String.class, tlsVersions, protocolsToSelectFrom);
-    return new Builder(this)
-        .cipherSuites(cipherSuitesToEnable)
-        .tlsVersions(tlsVersionsToEnable)
-        .build();
+    // We use enabled protocols here, not supported, to avoid re-enabling a protocol that has
+    // been disabled. Just because something is supported does not make it desirable to use.
+    String[] enabledProtocols = socket.getEnabledProtocols();
+    // The protocols must contain the leading protocol from the connection spec to be considered
+    // compatible. e.g. a connection spec containing [TLSv1.2, TLSv1.1, etc.] is only compatible
+    // if the socket has TLSv1.2 enabled.
+    String requiredProtocol = tlsVersions[0];
+    boolean requiredProtocolEnabled = contains(enabledProtocols, requiredProtocol);
+
+    boolean requiredCiphersEnabled;
+    if (cipherSuites == null) {
+      requiredCiphersEnabled = socket.getEnabledCipherSuites().length > 0;
+    } else {
+      String[] enabledCipherSuites = socket.getEnabledCipherSuites();
+      requiredCiphersEnabled = false;
+      for (String requiredCipher : cipherSuites) {
+        if (contains(enabledCipherSuites, requiredCipher)) {
+          requiredCiphersEnabled = true;
+          break;
+        }
+      }
+    }
+    return requiredProtocolEnabled && requiredCiphersEnabled;
+  }
+
+  private static <T> boolean contains(T[] array, T value) {
```

Comment:
AFAIK, no. The first arg comes from getEnabledProtocols() and getEnabledCipherSuites() which should always return an array, even if empty. I can make it more defensive though because this could be interpreted differently by different implementers of SSLSocket.